doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,257 @@
+    collection_id: CollectionType,
+}
+
+/// Specifies the type of a collection associated with a source.
+enum CollectionType {
+    /// The collection containing the reclocked upstream data.
+    Reclocked,
+    /// The collection containing the reclocking mappings.
+    Remapping,
+}
+```
+
+### SQL changes
+
+The `CREATE SOURCE` statement will be extended with an `EXPOSE RECLOCKING
+AS <name>` clause. For example:
+
+```sql
+CREATE SOURCE foo
+FROM KAFKA BROKER '...'
+EXPOSE RECLOCKING AS my_foo_clock
+```
+
+Materialize will make the contents of the `remap` collection for the source
+visible under the specified name via a relation whose structure depends on the
+type of the source:
+
+* File sources
+
+  | mz_timestamp | offset |
+  |--------------|--------|
+  | ...          | ...    |
+
+* Kafka sources
+
+  | mz_timestamp | partition | offset |
+  |--------------|-----------|--------|
+  | ...          | ...       | ...    |
+
+* Kinesis sources
+
+  | mz_timestamp | shard_id | sequence_number |
+  |--------------|----------|-----------------|
+  | ...          | ...      | ...             |
+
+* PostgreSQL sources
+
+  | mz_timestamp | lsn |
+  |--------------|-----|
+  | ...          | ... |
+
+* S3 sources
+
+  | mz_timestamp | key | offset |
+  |--------------|-----|--------|
+  | ...          | ... | ...    |
+
+
+If `EXPOSE RECLOCKING` is not specified, and the source type is reclockable,
+Materialize will generate a name of the form `<source>_reclocking`. If that name
+already exists, Materialize will try names of the form `<source>_reclocking1`,
+`<source>_reclocking2`, ... in sequence until it finds an available name. This
+matches the logic used for generatic the name of a primary index for a table.
+
+The reclocked relation will be a new catalog item of type `Reclocking`:
+
+```rust
+// catalog.rs
+/// An item representing a source reclocking relation.
+struct Reclocking {
+    /// The ID of the source to which this reclocking applies.
+    source_id: GlobalId,
+    /// The type of the reclocking relation.
+    ///
+    /// This must be derived the type of the `source_id` source. It is
+    /// duplicated here, rather than being derived from the source on the fly,
+    /// due to limitations in the design of the catalog API. (The type of the
+    /// item must be derivable from the item itself, i.e., without access to the
+    /// rest of catalog).
+    desc: RelationDesc,
+}
+```
+
+`Reclocking` items will necessarily have a `GlobalId` in order to be included in
+the catalog; however this `GlobalId` will not be meaningful to the storage
+layer. (TODO: should it be? Should we instead have `CreateSources` accept
+separate `GlobalId`s for the remap and reclocked collections, for the source
+types that support both? We'd need to ensure that `Insert, `AllowCompaction`,
+et al. are not supported for remap collections.)
+
+### System catalog changes
+
+A new system table `mz_reclockings` will be added with the following schema:
```

Comment:
Ha! I almost DM'd you to ask about this. I initially wrote this up to have `mz_kafka_reclockings` and friends. And then I realized... I had no idea how to implement that. Do you? Would be swell, if so, because I think it is a nicer end user experience to just have one of these per source type.

But, the problem to me seemed to be that combining these into one table means that you can't observe a timestamp binding until _all_ sources of the type have produced a binding at that timestamp. E.g., if ts 42 is assigned by source 1, you won't be able to see that in the mapping until ts 42 is assigned by sources 2-5. And even if we decide that's acceptable, how do we concretely produce such a collection? Do you drop and recreate the `mz_kafka_reclockings` dataflow every time a new source is added?

Reply:
It's a great point that it essentially stalls out all other sources, and I retract my interest in it!

Though, I think you could get around not having a `mz_kafka_reclockings` *dataflow* and just have it be a source that STORAGE can populate?

Reply:
> Though, I think you could get around not having a `mz_kafka_reclockings` _dataflow_ and just have it be a source that STORAGE can populate?

Yeah, that sounds right; you could just tell STORAGE to expose five built-in reclocking collections, one for each source type. 

But yeah, seems like a blast radius disaster! That one slow Kafka source will stall out all the other Kafka sources!

Reply:
The blast radius .. could just be part of the spec. STORAGE's job with reclocking is to emit bindings for the current timestamp ASAP. If it doesn't have new offsets yet, they aren't part of the next timestamp, and time moves forward without them. But .. yeah, maybe there needs to be a good argument for coalescing all the data. Might one want to query it? Like, "I have a fleet of sources, and want to know how up to date all of them are"?

Reply:
> Might one want to query it? Like, "I have a fleet of sources, and want to know how up to date all of them are"?

One very well might! This is a problem that comes up in general, though, if you have dynamically created _n_ sources or views in Materialize with identical schemas and you want to union them all together. Our answer right now is... use dbt to generate a view like the following:

```
CREATE VIEW all_my_kafkas AS SELECT * FROM kafka1_reclocking UNION ALL SELECT * FROM kafka2_reclocking;
```

That's probably a long-term crappy answer, but it does the trick in the short term. But it seems to me there is a general problem here of "I'd like to union together a dynamic set of relations with the same schema", and perhaps we just seek to solve that general problem down the road.